task_manager/task_manager.hh

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include "replica/database.hh"
+
+using task_id = utils::tagged_uuid<struct task_id_tag>;
+using is_abortable = bool_class <struct abortable_tag>;
+
+class task_manager : public seastar::peering_sharded_service<task_manager> {
+public:
+    class task;
+    using task_ptr = seastar::shared_ptr<task_manager::task>;
+    using task_map = std::unordered_map<task_id, task_ptr>;
+    using foreign_task_ptr = seastar::foreign_ptr<task_ptr>;
+    using foreign_task_vector = std::vector<foreign_task_ptr>;
+    using tasks_per_module = std::map<std::string, task_map>;
+    using modules = std::vector<std::string>;
+private:
+    tasks_per_module _tasks_per_module;
+    task_map _all_tasks;
+    modules _modules;
+    seastar::gate _gate;
+    const db::config& _cfg;
+    abort_source& _as;
+public:
+    enum class task_state {
+        created,
+        running,
+        done,
+        failed
+    };
+    class task {
+    public:
+        struct progress {
+            double completed = 0.0;         // Number of units completed so far.
+            double total = 0.0;             // Total number of units to complete the task.
+        };
+        struct status {
+            task_id id;
+            std::string type;
+            task_state state = task_state::created;
+            db_clock::time_point start_time;
+            db_clock::time_point end_time;
+            std::string error;
+            int number = 0;                 // A running sequence number of the task.
+            unsigned shard = 0;
+            std::string keyspace;
+            std::string table;
+            std::string entity;             // Additional entity specific for the given type of task.
+            std::string progress_units;     // A description of the units progress.
+            progress progress;
+        };
+     protected:
+        status _status;
+        std::unique_ptr<foreign_task_ptr> _parent = nullptr;
+        foreign_task_vector _children;
+        seastar::shared_promise<> _done;
+    public:
+        task(task_id id, foreign_task_ptr&& parent = nullptr) {
+            _status.id = id;
+            _status.state = task_state::created;
+            _status.shard = this_shard_id();
+            if (parent.get() != nullptr) {
+                _parent = std::make_unique<foreign_task_ptr>(std::move(parent));
+            }
+        }
+
+        status& get_status() {
+            return _status;
+        }
+
+        bool has_parent() const {
+            return _parent != nullptr;
+        }
+
+        // Can be called only if the task has a parent task.
```

Comment:
It can just return a null uuid otherwise (`task_id::create_null_id()`).
It is invalid for a real task id and can be tested as a boolean to signify that that task has no parent.